nexus/src/db/queries/region_allocation.rs

```diff
@@ -0,0 +1,540 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+//! Implementation of queries for provisioning regions.
+
+use crate::db::alias::ExpressionAlias;
+use crate::db::datastore::REGION_REDUNDANCY_THRESHOLD;
+use crate::db::model::{DatasetKind, Region};
+use crate::db::pool::DbConnection;
+use crate::db::subquery::{
+    AsQuerySource, Cte, CteBuilder, CteQuery, TrueOrCastError,
+};
+use db_macros::Subquery;
+use diesel::pg::Pg;
+use diesel::query_builder::{AstPass, Query, QueryFragment, QueryId};
+use diesel::{
+    sql_types, BoolExpressionMethods, Column, CombineDsl, ExpressionMethods,
+    Insertable, IntoSql, JoinOnDsl, NullableExpressionMethods, QueryDsl,
+    RunQueryDsl,
+};
+use nexus_db_model::queries::region_allocation::{
+    candidate_datasets, candidate_regions, candidate_zpools, do_insert,
+    inserted_regions, old_regions, old_zpool_usage, proposed_dataset_changes,
+    proposed_datasets_fit, updated_datasets, zpool_size_delta,
+};
+use nexus_db_model::schema;
+use nexus_db_model::ByteCount;
+use omicron_common::api::external;
+
+const NOT_ENOUGH_DATASETS_SENTINEL: &'static str = "Not enough datasets";
+const NOT_ENOUGH_ZPOOL_SPACE_SENTINEL: &'static str = "Not enough space";
+
+fn bool_parse_error(sentinel: &'static str) -> String {
+    format!("could not parse \"{sentinel}\" as type bool: invalid bool value")
+}
+
+/// Translates a generic pool error to an external error based
+/// on messages which may be emitted during region provisioning.
+pub fn from_pool(e: async_bb8_diesel::PoolError) -> external::Error {
+    use crate::db::error;
+    use async_bb8_diesel::ConnectionError;
+    use async_bb8_diesel::PoolError;
+    use diesel::result::DatabaseErrorKind;
+    use diesel::result::Error;
+
+    match e {
+        // Catch the specific errors designed to communicate the failures we
+        // want to distinguish
+        PoolError::Connection(ConnectionError::Query(
+            Error::DatabaseError(DatabaseErrorKind::Unknown, ref info),
+        )) => {
+            if info.message() == bool_parse_error(NOT_ENOUGH_DATASETS_SENTINEL)
+            {
+                return external::Error::unavail(
+                    "Not enough datasets to allocate disks",
+                );
+            } else if info.message()
+                == bool_parse_error(NOT_ENOUGH_ZPOOL_SPACE_SENTINEL)
+            {
+                return external::Error::unavail(
+                    "Not enough zpool space to allocate disks",
+                );
+            }
+        }
+        // Any other error at all is a bug
+        _ => {}
```

Comment:
I intentionally fell-through to avoid repeating the conversion multiple times - if I do the error conversion in this match arm, I don't have a conversion for the "sentinel does not match `info.message()` case (unless I duplicate it there too).